calico/felix/labels.py

```diff
@@ -0,0 +1,480 @@
+# -*- coding: utf-8 -*-
+# Copyright 2016 Metaswitch Networks
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+
+from calico.calcollections import MultiDict
+from calico.felix.selectors import LabelToLiteralEquality, LabelInSetLiteral
+
+_log = logging.getLogger(__name__)
+
+
+class LinearScanLabelIndex(object):
+    """
+    A label index matches a set of SelectorExpressions against a set of
+    label dicts.  As the matches between the two collections change,
+    it triggers calls to on_match_started/on_match_stopped.
+
+    A simple implementation of a label index.  Every update is handled by
+    a full linear scan.
+
+    This class has two purposes:
+
+    - it provides a benchmark against which other implementations can be
+      measured
+    - since it's simple, it's useful for comparative testing; any
+      other label index implementation should give the same end result.
```

Comment:
# Also, apparently, it's a base class for other more advanced
# implementations.  (Although I've not yet seen whether any of this
# class's methods are actually used by subclasses.)